editor/editor_dock_manager.cpp

```diff
@@ -0,0 +1,712 @@
+				dock_slot[atidx]->set_block_signals(true);
+				dock_slot[i]->move_tab_from_tab_container(dock_slot[atidx], dock_slot[atidx]->get_tab_idx_from_control(node), 0);
+				dock_slot[i]->set_block_signals(false);
+				dock_slot[atidx]->set_block_signals(false);
+			}
+
+			WindowWrapper *wrapper = Object::cast_to<WindowWrapper>(node);
+			if (restore_window_on_load && floating_docks_dump.has(name)) {
+				if (!dock_slot[i]->is_tab_hidden(dock_slot[i]->get_tab_idx_from_control(node))) {
+					_restore_floating_dock(floating_docks_dump[name], node, i);
+				}
+			} else if (wrapper) {
+				wrapper->set_window_enabled(false);
+			}
+		}
+
+		if (!p_layout->has_section_key(p_section, "dock_" + itos(i + 1) + "_selected_tab_idx")) {
+			continue;
+		}
+
+		int selected_tab_idx = p_layout->get_value(p_section, "dock_" + itos(i + 1) + "_selected_tab_idx");
+		if (selected_tab_idx >= 0 && selected_tab_idx < dock_slot[i]->get_tab_count()) {
+			callable_mp(dock_slot[i], &TabContainer::set_current_tab).call_deferred(selected_tab_idx);
+		}
+	}
+
+	for (int i = 0; i < vsplits.size(); i++) {
+		if (!p_layout->has_section_key(p_section, "dock_split_" + itos(i + 1))) {
+			continue;
+		}
+
+		int ofs = p_layout->get_value(p_section, "dock_split_" + itos(i + 1));
+		vsplits[i]->set_split_offset(ofs);
+	}
+
+	for (int i = 0; i < hsplits.size(); i++) {
+		if (!p_layout->has_section_key(p_section, "dock_hsplit_" + itos(i + 1))) {
+			continue;
+		}
+		int ofs = p_layout->get_value(p_section, "dock_hsplit_" + itos(i + 1));
+		hsplits[i]->set_split_offset(ofs);
+	}
+
+	update_dock_slots_visibility(false);
+
+	FileSystemDock::get_singleton()->load_layout_from_config(p_layout, p_section);
+}
+
+void EditorDockManager::update_dock_slots_visibility(bool p_keep_selected_tabs) {
+	if (!docks_visible) {
+		for (int i = 0; i < DOCK_SLOT_MAX; i++) {
+			dock_slot[i]->hide();
+		}
+	} else {
+		for (int i = 0; i < DOCK_SLOT_MAX; i++) {
+			int first_tab_visible = -1;
+			for (int j = 0; j < dock_slot[i]->get_tab_count(); j++) {
+				if (!dock_slot[i]->is_tab_hidden(j)) {
+					first_tab_visible = j;
+					break;
+				}
+			}
+			if (first_tab_visible >= 0) {
+				dock_slot[i]->show();
+				if (p_keep_selected_tabs) {
+					int current_tab = dock_slot[i]->get_current_tab();
+					if (dock_slot[i]->is_tab_hidden(current_tab)) {
+						dock_slot[i]->set_block_signals(true);
+						dock_slot[i]->select_next_available();
+						dock_slot[i]->set_block_signals(false);
+					}
+				} else {
+					dock_slot[i]->set_block_signals(true);
+					dock_slot[i]->set_current_tab(first_tab_visible);
+					dock_slot[i]->set_block_signals(false);
+				}
+			} else {
+				dock_slot[i]->hide();
+			}
+		}
+	}
+}
+
+void EditorDockManager::close_all_floating_docks() {
+	for (WindowWrapper *wrapper : floating_docks) {
+		wrapper->set_window_enabled(false);
+	}
+}
+
+void EditorDockManager::add_control_to_dock(DockSlot p_slot, Control *p_control, const String &p_name) {
```

Comment:
Missing `update_dock_slots_visibility()` at the end, isn't it?

Reply:
Adding the control to the dock calls `DockSplitContainer::add_child_notify()` which updates its and its parent DockSplitContainer's visibility.

Reply:
I considered that, but then the remove method is wrong, because `remove_child_notify` also should update the visibility.

Reply:
Its because when removing a control, the current tab index changes and there could be only a hidden tab left, so `update_dock_slots_visibility` is necessary, since DockSplitContainer doesn't deal with TabContainer tabs.
When adding a tab, it is guaranteed to not be hidden and the current tab won't change.

I could add the `update_dock_slots_visibility()` anyway, if we want.

Reply:
Alright, I couldn't find any issues with the behavior anyway. So let's consider this correct for now. Thanks for the explanation!